--- a/usr/src/nextcloud/lib/private/Preview/TXT.php
+++ b/usr/src/nextcloud/lib/private/Preview/TXT.php
@@ -76,9 +76,7 @@
                imagecolorallocate($image, 255, 255, 255);
                $textColor = imagecolorallocate($image, 0, 0, 0);

-               $fontFile = __DIR__;
-               $fontFile .= '/../../../core';
-               $fontFile .= '/fonts/NotoSans-Regular.ttf';
+               $fontFile = '/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc';

                $canUseTTF = function_exists('imagettftext');
